File: src/checkins.js
const { usageError } = require('./errors');
const { parseDateString } = require('./date');

function findCheckinIndex(db, habitId, date) {
  return db.checkins.findIndex((c) => c.habit_id === habitId && c.date === date);
}

function getQuantity(db, habitId, date) {
  const idx = findCheckinIndex(db, habitId, date);
  if (idx === -1) return 0;
  return db.checkins[idx].quantity;
}

function setQuantity(db, habitId, date, quantity) {
  parseDateString(date);
  if (!Number.isInteger(quantity) || quantity < 0) throw usageError('Invalid quantity');

  const idx = findCheckinIndex(db, habitId, date);
  if (quantity === 0) {
    if (idx !== -1) db.checkins.splice(idx, 1);
    return;
  }
  if (idx === -1) {
    db.checkins.push({ habit_id: habitId, date, quantity });
  } else {
    db.checkins[idx].quantity = quantity;
  }
}

function addQuantity(db, habitId, date, delta) {
  parseDateString(date);
  if (!Number.isInteger(delta) || delta < 1) throw usageError('Invalid quantity');
  const cur = getQuantity(db, habitId, date);
  setQuantity(db, habitId, date, cur + delta);
  return cur + delta;
}

function listCheckinsForHabit(db, habitId) {
  return db.checkins
    .filter((c) => c.habit_id === habitId)
    .slice()
    .sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : (a.habit_id < b.habit_id ? -1 : a.habit_id > b.habit_id ? 1 : 0)));
}

function listCheckinsInRange(db, { from, to, habitIds }) {
  return db.checkins
    .filter((c) => {
      if (habitIds && !habitIds.has(c.habit_id)) return false;
      if (from && c.date < from) return false;
      if (to && c.date > to) return false;
      return true;
    })
    .slice()
    .sort((a, b) => (a.date < b.date ? -1 : a.date > b.date ? 1 : (a.habit_id < b.habit_id ? -1 : a.habit_id > b.habit_id ? 1 : 0)));
}

module.exports = {
  getQuantity,
  setQuantity,
  addQuantity,
  listCheckinsForHabit,
  listCheckinsInRange
};

File: src/checkins.rs
use crate::date::parse_date_string;
use crate::error::CliError;
use crate::model::{Checkin, Db};
use std::collections::BTreeSet;

fn find_checkin_index(db: &Db, habit_id: &str, date: &str) -> Option<usize> {
    db.checkins
        .iter()
        .position(|c| c.habit_id == habit_id && c.date == date)
}

pub fn get_quantity(db: &Db, habit_id: &str, date: &str) -> u32 {
    match find_checkin_index(db, habit_id, date) {
        Some(i) => db.checkins[i].quantity,
        None => 0,
    }
}

pub fn set_quantity(db: &mut Db, habit_id: &str, date: &str, quantity: u32) -> Result<(), CliError> {
    parse_date_string(date, "date")?;

    let idx = find_checkin_index(db, habit_id, date);

    if quantity == 0 {
        if let Some(i) = idx {
            db.checkins.remove(i);
        }
        return Ok(());
    }

    match idx {
        None => db.checkins.push(Checkin {
            habit_id: habit_id.to_string(),
            date: date.to_string(),
            quantity,
        }),
        Some(i) => db.checkins[i].quantity = quantity,
    }

    Ok(())
}

pub fn add_quantity(db: &mut Db, habit_id: &str, date: &str, delta: u32) -> Result<u32, CliError> {
    parse_date_string(date, "date")?;
    if delta < 1 {
        return Err(CliError::usage("Invalid quantity"));
    }

    let cur = get_quantity(db, habit_id, date);
    let total = cur.saturating_add(delta);
    set_quantity(db, habit_id, date, total)?;
    Ok(total)
}

pub fn list_checkins_for_habit(db: &Db, habit_id: &str) -> Vec<Checkin> {
    let mut out: Vec<Checkin> = db
        .checkins
        .iter()
        .filter(|c| c.habit_id == habit_id)
        .cloned()
        .collect();

    out.sort_by(|a, b| {
        if a.date != b.date {
            a.date.cmp(&b.date)
        } else {
            a.habit_id.cmp(&b.habit_id)
        }
    });

    out
}

pub fn list_checkins_in_range(
    db: &Db,
    from: Option<&str>,
    to: Option<&str>,
    habit_ids: Option<&BTreeSet<String>>,
) -> Vec<Checkin> {
    let mut out: Vec<Checkin> = db
        .checkins
        .iter()
        .filter(|c| {
            if let Some(ids) = habit_ids {
                if !ids.contains(&c.habit_id) {
                    return false;
                }
            }
            if let Some(f) = from {
                if c.date < f {
                    return false;
                }
            }
            if let Some(t) = to {
                if c.date > t {
                    return false;
                }
            }
            true
        })
        .cloned()
        .collect();

    out.sort_by(|a, b| {
        if a.date != b.date {
            a.date.cmp(&b.date)
        } else {
            a.habit_id.cmp(&b.habit_id)
        }
    });

    out
}
